aot.py

```diff
@@ -0,0 +1,65 @@
+import os
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+
+from functorch.compile import aot_function
+from functorch.compile import clear_compile_cache
+
+def setup(rank, world_size):
+    os.environ['MASTER_ADDR'] = 'localhost'
+    os.environ['MASTER_PORT'] = '12355'
+
+    # initialize the process group
+    dist.init_process_group("nccl", rank=rank, world_size=world_size)
+
+
+def cleanup():
+    dist.destroy_process_group()
+
+def broadcast():
+    device = torch.device("cuda", dist.get_rank())
+    x = torch.zeros(2, 3).to(device)
+    if device.index == 0:
+        x = torch.ones(2, 3).to(device)
+    dist.broadcast(x, 0)
+    print(f"{os.getpid()} broadcast: {x.cpu()}")
+    return x
+
+# The compiler_fn is called after the forward and backward graphs are extracted.
+# Here, we just print the code in the compiler_fn. Return of this function is a callable.
+def compiler_fn(fx_module: torch.fx.GraphModule, _):
+    print(fx_module.code)
+    return fx_module
+
+def demo_basic(rank, world_size):
+    # without flushing, mp won't print this message if crash.
+    print(f"Running basic DDP example on rank {rank}, and process id: {os.getpid()}", flush=True)
+    setup(rank, world_size)
+
+    # Pass on the compiler_fn to the aot_function API
+    aot_print_fn = aot_function(broadcast, fw_compiler=compiler_fn, bw_compiler=compiler_fn)
```

Comment:
@Chillee I added example of AOT here. And this is the output. Looks like it works just don't know how to deal with the `bw_compiler`:
```
(pytorch37) jwtan@dev-st-p4d24xlarge-2:/fsx/users/jwtan/work/pytorch$ python aot.py 
main process id: 96435
Running basic DDP example on rank 0, and process id: 96478
Running basic DDP example on rank 1, and process id: 96479
96478 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:0')
96479 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:1')
96478 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:0')
96479 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:1')



def forward(self):
    _tensor_constant0 = self._tensor_constant0
    return [_tensor_constant0]
    



def forward(self, tangents_1):
    return []
    



def forward(self):
    _tensor_constant0 = self._tensor_constant0
    return [_tensor_constant0]
    



def forward(self, tangents_1):
    return []
    
96479 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:1')
96478 broadcast: tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:0')
```

Reply:
Looks like AOT doesn't work... Debugging...

Reply:
Added a print after __torch_dispatch__ in AOT: https://github.com/pytorch/functorch/blob/main/functorch/_src/python_key.py#L73

Nothing got printed out... So it seems __torch_dispatch__ cannot recognize torch.ops...